File: android/src/main/kotlin/com/polyfence/polyfence/utils/PolyfenceConfig.kt
package com.polyfence.polyfence.utils

import android.content.Context
import android.content.SharedPreferences
import android.util.Log

/**
 * Centralized configuration management for Polyfence
 * Single responsibility: Runtime configuration and persistence
 */
class PolyfenceConfig(context: Context) {
    
    companion object {
        private const val TAG = "PolyfenceConfig"
        private const val PREFS_NAME = "polyfence_config"
        
        // Default GPS Configuration
        const val DEFAULT_GPS_INTERVAL_MS = 5000L // 5 seconds - balanced
        const val DEFAULT_GPS_ACCURACY_THRESHOLD = 100.0f
        const val DEFAULT_MIN_UPDATE_INTERVAL_MS = 1000L
        const val DEFAULT_MAX_UPDATE_DELAY_MS = 6000L
        const val MIN_UPDATE_DISTANCE_METERS = 10f
        
        // Zone Validation Configuration
        const val DEFAULT_CONFIDENCE_POINTS = 2
        const val DEFAULT_CONFIDENCE_TIMEOUT_MS = 10000L
        const val DEFAULT_REQUIRE_CONFIRMATION = true
        const val LARGE_ZONE_RADIUS_THRESHOLD_METERS = 200.0
        const val MIN_SINGLE_POINT_ZONE_RADIUS_METERS = 50.0
        const val MIN_POLYGON_POINTS = 3
        
        // Speed and Movement Thresholds
        const val HIGH_SPEED_THRESHOLD_KMH = 40.0
        const val SPEED_MS_TO_KMH_MULTIPLIER = 3.6
        
        // GPS Health and Recovery
        const val MAX_GPS_FAILURES_BEFORE_COOLDOWN = 2
        const val MIN_GPS_FAILURES_FOR_RECOVERY = 3
        const val MAX_GPS_FAILURES_FOR_RECOVERY = 5
        const val GPS_HEALTH_CHECK_TIMEOUT_MS = 120000L // 2 minutes
        const val HEALTH_CHECK_INTERVAL_MS = 30000L // 30 seconds
        const val GPS_RESTART_DELAY_MS = 3000L // 3 seconds
        const val SERVICE_RESTART_DELAY_MS = 5000L // 5 seconds
        
        // GPS Restart Configuration
        const val MIN_GPS_RESTART_INTERVAL_MS = 10000L // 10 seconds
        const val MIN_UPDATE_RESTART_INTERVAL_MS = 5000L // 5 seconds
        const val MAX_UPDATE_RESTART_DELAY_MS = 15000L // 15 seconds
        
        // System Defaults
        const val DEFAULT_BATTERY_LEVEL = 100.0
        const val DEVICE_ID_RANDOM_RANGE = 10000
        const val FOREGROUND_NOTIFICATION_ID = 1001
        const val APP_VERSION = "1.0.0"
        
        // Validation Ranges
        const val MIN_GPS_INTERVAL_MS = 1000L
        const val MAX_GPS_INTERVAL_MS = 60000L
        const val MIN_ACCURACY_THRESHOLD = 1.0f
        const val MAX_ACCURACY_THRESHOLD = 500.0f
        const val MIN_CONFIDENCE_POINTS = 1
        const val MAX_CONFIDENCE_POINTS = 5
        
        // Calculation Factors
        const val MIN_UPDATE_INTERVAL_FACTOR = 2L // minInterval = gpsInterval / 2
        const val MAX_UPDATE_DELAY_FACTOR = 2L // maxDelay = gpsInterval * 2
        
        // Cache Configuration
        const val DEFAULT_LRU_INITIAL_CAPACITY = 16
        const val DEFAULT_LRU_LOAD_FACTOR = 0.75f
    }
    
    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    
    // GPS Configuration
    var gpsIntervalMs: Long
        get() = prefs.getLong("gps_interval_ms", DEFAULT_GPS_INTERVAL_MS)
        set(value) {
            prefs.edit().putLong("gps_interval_ms", value).apply()
        }
    
    var gpsAccuracyThreshold: Float
        get() = prefs.getFloat("gps_accuracy_threshold", DEFAULT_GPS_ACCURACY_THRESHOLD)
        set(value) {
            prefs.edit().putFloat("gps_accuracy_threshold", value).apply()
        }
    
    var minUpdateIntervalMs: Long
        get() = prefs.getLong("min_update_interval_ms", DEFAULT_MIN_UPDATE_INTERVAL_MS)
        set(value) {
            prefs.edit().putLong("min_update_interval_ms", value).apply()
        }
    
    var maxUpdateDelayMs: Long
        get() = prefs.getLong("max_update_delay_ms", DEFAULT_MAX_UPDATE_DELAY_MS)
        set(value) {
            prefs.edit().putLong("max_update_delay_ms", value).apply()
        }
    
    // Validation Configuration
    var requireConfirmation: Boolean
        get() = prefs.getBoolean("require_confirmation", DEFAULT_REQUIRE_CONFIRMATION)
        set(value) {
            prefs.edit().putBoolean("require_confirmation", value).apply()
        }
    
    var confidencePoints: Int
        get() = prefs.getInt("confidence_points", DEFAULT_CONFIDENCE_POINTS)
        set(value) {
            prefs.edit().putInt("confidence_points", value).apply()
        }
    
    var confidenceTimeoutMs: Long
        get() = prefs.getLong("confidence_timeout_ms", DEFAULT_CONFIDENCE_TIMEOUT_MS)
        set(value) {
            prefs.edit().putLong("confidence_timeout_ms", value).apply()
        }
    
    /**
     * Reset all configuration to defaults
     */
    fun resetToDefaults() {
        prefs.edit().clear().apply()
    }
    
    /**
     * Get current configuration as a map for debugging
     */
    fun getConfigurationMap(): Map<String, Any> {
        return mapOf(
            "gps_interval_ms" to gpsIntervalMs,
            "gps_accuracy_threshold" to gpsAccuracyThreshold,
            "min_update_interval_ms" to minUpdateIntervalMs,
            "max_update_delay_ms" to maxUpdateDelayMs,
            "require_confirmation" to requireConfirmation,
            "confidence_points" to confidencePoints,
            "confidence_timeout_ms" to confidenceTimeoutMs
        )
    }
    
    /**
     * Update configuration from a map (useful for Flutter integration)
     */
    fun updateFromMap(configMap: Map<String, Any>) {
        try {
            configMap["gps_interval_ms"]?.let { 
                if (it is Number) gpsIntervalMs = it.toLong() 
            }
            configMap["gps_accuracy_threshold"]?.let { 
                if (it is Number) gpsAccuracyThreshold = it.toFloat() 
            }
            configMap["min_update_interval_ms"]?.let { 
                if (it is Number) minUpdateIntervalMs = it.toLong() 
            }
            configMap["max_update_delay_ms"]?.let { 
                if (it is Number) maxUpdateDelayMs = it.toLong() 
            }
            configMap["require_confirmation"]?.let { 
                if (it is Boolean) requireConfirmation = it 
            }
            configMap["confidence_points"]?.let { 
                if (it is Number) confidencePoints = it.toInt() 
            }
            configMap["confidence_timeout_ms"]?.let { 
                if (it is Number) confidenceTimeoutMs = it.toLong() 
            }
            
        } catch (e: Exception) {
            Log.e(TAG, "Failed to update configuration: ${e.message}")
        }
    }
    
    /**
     * Validate configuration values
     */
    fun validateAndCorrect(): Boolean {
        var corrected = false
        
        // Ensure GPS interval is reasonable
        if (gpsIntervalMs < MIN_GPS_INTERVAL_MS || gpsIntervalMs > MAX_GPS_INTERVAL_MS) {
            gpsIntervalMs = DEFAULT_GPS_INTERVAL_MS
            corrected = true
        }
        
        // Ensure accuracy threshold is reasonable
        if (gpsAccuracyThreshold < MIN_ACCURACY_THRESHOLD || gpsAccuracyThreshold > MAX_ACCURACY_THRESHOLD) {
            gpsAccuracyThreshold = DEFAULT_GPS_ACCURACY_THRESHOLD
            corrected = true
        }
        
        // Ensure confidence points is reasonable
        if (confidencePoints < MIN_CONFIDENCE_POINTS || confidencePoints > MAX_CONFIDENCE_POINTS) {
            confidencePoints = DEFAULT_CONFIDENCE_POINTS
            corrected = true
        }
        
        // Ensure min interval is less than main interval
        if (minUpdateIntervalMs >= gpsIntervalMs) {
            minUpdateIntervalMs = gpsIntervalMs / MIN_UPDATE_INTERVAL_FACTOR
            corrected = true
        }
        
        // Ensure max delay is greater than main interval
        if (maxUpdateDelayMs <= gpsIntervalMs) {
            maxUpdateDelayMs = gpsIntervalMs * MAX_UPDATE_DELAY_FACTOR
            corrected = true
        }
        
        if (corrected) {
            Log.w(TAG, "Configuration values corrected")
        }
        
        return !corrected
    }
    
    /**
     * Log current configuration for debugging
     */
    fun logCurrentConfig() {
    }
}

File: ios/Classes/Utils/PolyfenceConfig.swift
import Foundation

/**
 * Centralized configuration management for Polyfence iOS
 * Single responsibility: Runtime configuration and persistence
 * Ported from Android PolyfenceConfig.kt
 */
class PolyfenceConfig {
    
    // MARK: - Constants
    private static let TAG = "PolyfenceConfig"
    private static let PREFS_NAME = "polyfence_config"
    
    // Default GPS Configuration
    static let DEFAULT_GPS_INTERVAL_MS: TimeInterval = 5.0 // 5 seconds - balanced
    static let DEFAULT_GPS_ACCURACY_THRESHOLD: Double = 100.0
    static let DEFAULT_MIN_UPDATE_INTERVAL_MS: TimeInterval = 1.0
    static let DEFAULT_MAX_UPDATE_DELAY_MS: TimeInterval = 6.0
    static let MIN_UPDATE_DISTANCE_METERS: Double = 10.0
    
    // Zone Validation Configuration
    static let DEFAULT_CONFIDENCE_POINTS: Int = 2
    static let DEFAULT_CONFIDENCE_TIMEOUT_MS: TimeInterval = 10.0
    static let DEFAULT_REQUIRE_CONFIRMATION: Bool = true
    static let LARGE_ZONE_RADIUS_THRESHOLD_METERS: Double = 200.0
    static let MIN_SINGLE_POINT_ZONE_RADIUS_METERS: Double = 50.0
    static let MIN_POLYGON_POINTS: Int = 3
    
    // Speed and Movement Thresholds
    static let HIGH_SPEED_THRESHOLD_KMH: Double = 40.0
    static let SPEED_MS_TO_KMH_MULTIPLIER: Double = 3.6
    
    // GPS Health and Recovery
    static let MAX_GPS_FAILURES_BEFORE_COOLDOWN: Int = 2
    static let MIN_GPS_FAILURES_FOR_RECOVERY: Int = 3
    static let MAX_GPS_FAILURES_FOR_RECOVERY: Int = 5
    static let GPS_HEALTH_CHECK_TIMEOUT_MS: TimeInterval = 120.0 // 2 minutes
    static let HEALTH_CHECK_INTERVAL_MS: TimeInterval = 30.0 // 30 seconds
    static let GPS_RESTART_DELAY_MS: TimeInterval = 3.0 // 3 seconds
    static let SERVICE_RESTART_DELAY_MS: TimeInterval = 5.0 // 5 seconds
    
    // GPS Restart Configuration
    static let MIN_GPS_RESTART_INTERVAL_MS: TimeInterval = 10.0 // 10 seconds
    static let MIN_UPDATE_RESTART_INTERVAL_MS: TimeInterval = 5.0 // 5 seconds
    static let MAX_UPDATE_RESTART_DELAY_MS: TimeInterval = 15.0 // 15 seconds
    
    // System Defaults
    static let DEFAULT_BATTERY_LEVEL: Double = 100.0
    static let MOCK_CPU_USAGE_PERCENT: Double = 5.0
    static let DEVICE_ID_RANDOM_RANGE: Int = 10000
    static let FOREGROUND_NOTIFICATION_ID: Int = 1001
    static let APP_VERSION: String = "1.0.0"
    
    // Validation Ranges
    static let MIN_GPS_INTERVAL_MS: TimeInterval = 1.0
    static let MAX_GPS_INTERVAL_MS: TimeInterval = 60.0
    static let MIN_ACCURACY_THRESHOLD: Double = 1.0
    static let MAX_ACCURACY_THRESHOLD: Double = 500.0
    static let MIN_CONFIDENCE_POINTS: Int = 1
    static let MAX_CONFIDENCE_POINTS: Int = 5
    
    // Calculation Factors
    static let MIN_UPDATE_INTERVAL_FACTOR: TimeInterval = 2.0 // minInterval = gpsInterval / 2
    static let MAX_UPDATE_DELAY_FACTOR: TimeInterval = 2.0 // maxDelay = gpsInterval * 2
    
    // MARK: - Properties
    private let userDefaults = UserDefaults.standard
    
    // GPS Configuration
    var gpsIntervalMs: TimeInterval {
        get {
            return userDefaults.double(forKey: "gps_interval_ms")
        }
        set {
            userDefaults.set(newValue, forKey: "gps_interval_ms")
        }
    }
    
    var gpsAccuracyThreshold: Double {
        get {
            return userDefaults.double(forKey: "gps_accuracy_threshold")
        }
        set {
            userDefaults.set(newValue, forKey: "gps_accuracy_threshold")
        }
    }
    
    var minUpdateIntervalMs: TimeInterval {
        get {
            return userDefaults.double(forKey: "min_update_interval_ms")
        }
        set {
            userDefaults.set(newValue, forKey: "min_update_interval_ms")
        }
    }
    
    var maxUpdateDelayMs: TimeInterval {
        get {
            return userDefaults.double(forKey: "max_update_delay_ms")
        }
        set {
            userDefaults.set(newValue, forKey: "max_update_delay_ms")
        }
    }
    
    // Validation Configuration
    var requireConfirmation: Bool {
        get {
            return userDefaults.bool(forKey: "require_confirmation")
        }
        set {
            userDefaults.set(newValue, forKey: "require_confirmation")
        }
    }
    
    var confidencePoints: Int {
        get {
            return userDefaults.integer(forKey: "confidence_points")
        }
        set {
            userDefaults.set(newValue, forKey: "confidence_points")
        }
    }
    
    var confidenceTimeoutMs: TimeInterval {
        get {
            return userDefaults.double(forKey: "confidence_timeout_ms")
        }
        set {
            userDefaults.set(newValue, forKey: "confidence_timeout_ms")
        }
    }
    
    // MARK: - Initialization
    
    init() {
        // Set default values if not already set
        if userDefaults.object(forKey: "gps_interval_ms") == nil {
            gpsIntervalMs = PolyfenceConfig.DEFAULT_GPS_INTERVAL_MS
        }
        if userDefaults.object(forKey: "gps_accuracy_threshold") == nil {
            gpsAccuracyThreshold = PolyfenceConfig.DEFAULT_GPS_ACCURACY_THRESHOLD
        }
        if userDefaults.object(forKey: "min_update_interval_ms") == nil {
            minUpdateIntervalMs = PolyfenceConfig.DEFAULT_MIN_UPDATE_INTERVAL_MS
        }
        if userDefaults.object(forKey: "max_update_delay_ms") == nil {
            maxUpdateDelayMs = PolyfenceConfig.DEFAULT_MAX_UPDATE_DELAY_MS
        }
        if userDefaults.object(forKey: "require_confirmation") == nil {
            requireConfirmation = PolyfenceConfig.DEFAULT_REQUIRE_CONFIRMATION
        }
        if userDefaults.object(forKey: "confidence_points") == nil {
            confidencePoints = PolyfenceConfig.DEFAULT_CONFIDENCE_POINTS
        }
        if userDefaults.object(forKey: "confidence_timeout_ms") == nil {
            confidenceTimeoutMs = PolyfenceConfig.DEFAULT_CONFIDENCE_TIMEOUT_MS
        }
    }
    
    // MARK: - Public Methods
    
    /**
     * Validate and correct configuration values
     */
    func validateAndCorrect() {
        // Ensure GPS interval is reasonable
        if gpsIntervalMs < PolyfenceConfig.MIN_GPS_INTERVAL_MS || gpsIntervalMs > PolyfenceConfig.MAX_GPS_INTERVAL_MS {
            gpsIntervalMs = PolyfenceConfig.DEFAULT_GPS_INTERVAL_MS
        }
        
        // Ensure accuracy threshold is reasonable
        if gpsAccuracyThreshold < PolyfenceConfig.MIN_ACCURACY_THRESHOLD || gpsAccuracyThreshold > PolyfenceConfig.MAX_ACCURACY_THRESHOLD {
            gpsAccuracyThreshold = PolyfenceConfig.DEFAULT_GPS_ACCURACY_THRESHOLD
        }
        
        // Ensure confidence points are reasonable
        if confidencePoints < PolyfenceConfig.MIN_CONFIDENCE_POINTS || confidencePoints > PolyfenceConfig.MAX_CONFIDENCE_POINTS {
            confidencePoints = PolyfenceConfig.DEFAULT_CONFIDENCE_POINTS
        }
        
        // Ensure min update interval is less than GPS interval
        if minUpdateIntervalMs >= gpsIntervalMs {
            minUpdateIntervalMs = gpsIntervalMs / PolyfenceConfig.MIN_UPDATE_INTERVAL_FACTOR
        }
        
        // Ensure max update delay is greater than GPS interval
        if maxUpdateDelayMs <= gpsIntervalMs {
            maxUpdateDelayMs = gpsIntervalMs * PolyfenceConfig.MAX_UPDATE_DELAY_FACTOR
        }
        
        // Configuration validated successfully
    }
    
    /**
     * Get configuration as dictionary
     */
    func getConfiguration() -> [String: Any] {
        return [
            "gps_interval_ms": gpsIntervalMs,
            "gps_accuracy_threshold": gpsAccuracyThreshold,
            "min_update_interval_ms": minUpdateIntervalMs,
            "max_update_delay_ms": maxUpdateDelayMs,
            "require_confirmation": requireConfirmation,
            "confidence_points": confidencePoints,
            "confidence_timeout_ms": confidenceTimeoutMs
        ]
    }
    
    /**
     * Update configuration from dictionary
     */
    func updateConfiguration(_ configMap: [String: Any]) {
        if let gpsInterval = configMap["gps_interval_ms"] as? Double {
            gpsIntervalMs = gpsInterval
        }
        
        if let accuracyThreshold = configMap["gps_accuracy_threshold"] as? Double {
            gpsAccuracyThreshold = accuracyThreshold
        }
        
        if let minUpdateInterval = configMap["min_update_interval_ms"] as? Double {
            minUpdateIntervalMs = minUpdateInterval
        }
        
        if let maxUpdateDelay = configMap["max_update_delay_ms"] as? Double {
            maxUpdateDelayMs = maxUpdateDelay
        }
        
        if let requireConf = configMap["require_confirmation"] as? Bool {
            requireConfirmation = requireConf
        }
        
        if let confPoints = configMap["confidence_points"] as? Int {
            confidencePoints = confPoints
        }
        
        if let confTimeout = configMap["confidence_timeout_ms"] as? Double {
            confidenceTimeoutMs = confTimeout
        }
        
        // Validate after update
        validateAndCorrect()
        
        // Configuration updated successfully
    }
    
    /**
     * Reset configuration to defaults
     */
    func resetConfiguration() {
        gpsIntervalMs = PolyfenceConfig.DEFAULT_GPS_INTERVAL_MS
        gpsAccuracyThreshold = PolyfenceConfig.DEFAULT_GPS_ACCURACY_THRESHOLD
        minUpdateIntervalMs = PolyfenceConfig.DEFAULT_MIN_UPDATE_INTERVAL_MS
        maxUpdateDelayMs = PolyfenceConfig.DEFAULT_MAX_UPDATE_DELAY_MS
        requireConfirmation = PolyfenceConfig.DEFAULT_REQUIRE_CONFIRMATION
        confidencePoints = PolyfenceConfig.DEFAULT_CONFIDENCE_POINTS
        confidenceTimeoutMs = PolyfenceConfig.DEFAULT_CONFIDENCE_TIMEOUT_MS
        
        // Configuration reset successfully
    }
} 